component/common/loki/client/internal/marker_file_handler.go

```diff
@@ -0,0 +1,87 @@
+package internal
+
+import (
+	"fmt"
+	"github.com/grafana/agent/component/common/loki/wal"
+	"os"
+	"path/filepath"
+	"strconv"
+
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	"github.com/prometheus/prometheus/tsdb/fileutil"
+)
+
+type MarkerFileHandler interface {
+	wal.Marker
+	MarkSegment(segment int)
+}
+
+type markerFileHandler struct {
+	logger log.Logger
+
+	lastMarkedSegmentFilePath string
+}
+
+var (
+	_ MarkerFileHandler = (*markerFileHandler)(nil)
+)
+
+func NewMarkerFileHandler(logger log.Logger, walDir string) (MarkerFileHandler, error) {
+	markerDir := filepath.Join(walDir, "remote")
+	// attempt to create dir if doesn't exist
+	if err := os.MkdirAll(markerDir, 0o777); err != nil {
+		return nil, fmt.Errorf("error creating segment marker folder %q: %w", markerDir, err)
+	}
+
+	mfh := &markerFileHandler{
+		logger:                    logger,
+		lastMarkedSegmentFilePath: filepath.Join(markerDir, "segment"),
+	}
+
+	return mfh, nil
+}
+
+// LastMarkedSegment implements wlog.Marker.
+func (mfh *markerFileHandler) LastMarkedSegment() int {
+	bb, err := os.ReadFile(mfh.lastMarkedSegmentFilePath)
+	if os.IsNotExist(err) {
+		level.Warn(mfh.logger).Log("msg", "marker segment file does not exist", "file", mfh.lastMarkedSegmentFilePath)
+		return -1
+	} else if err != nil {
+		level.Error(mfh.logger).Log("msg", "could not access segment marker file", "file", mfh.lastMarkedSegmentFilePath, "err", err)
+		return -1
+	}
+
+	savedSegment, err := strconv.Atoi(string(bb))
+	if err != nil {
+		level.Error(mfh.logger).Log("msg", "could not read segment marker file", "file", mfh.lastMarkedSegmentFilePath, "err", err)
+		return -1
+	}
+
+	if savedSegment < 0 {
+		level.Error(mfh.logger).Log("msg", "invalid segment number inside marker file", "file", mfh.lastMarkedSegmentFilePath, "segment number", savedSegment)
+		return -1
+	}
+
+	return savedSegment
+}
+
+// MarkSegment implements MarkerHandler.
+func (mfh *markerFileHandler) MarkSegment(segment int) {
+	var (
+		segmentText = strconv.Itoa(segment)
+		tmp         = mfh.lastMarkedSegmentFilePath + ".tmp"
+	)
+
+	if err := os.WriteFile(tmp, []byte(segmentText), 0o666); err != nil {
+		level.Error(mfh.logger).Log("msg", "could not create segment marker file", "file", tmp, "err", err)
+		return
+	}
+	if err := fileutil.Replace(tmp, mfh.lastMarkedSegmentFilePath); err != nil {
```

Comment:
Ok, just did a read over atomicity in renaming files. Let me split the answer in:
1. atomic marker swap
2. marker corruption
3. appending

For 1), just found here that if we use `os.Rename` the operation is [atomic in POSIX OSes](https://pkg.go.dev/os#Rename), and in Windows, according to [this](https://stackoverflow.com/questions/30385225/is-there-an-os-independent-way-to-atomically-overwrite-a-file), should be atomic if the source and destination are the same media (which they are in our case). I think we could use `os.Rename` and `os.CreateTemp` to implement the "atomic write" of the marker. In the Windows case, I think we could have a "this should work" for now, and then with some more digging and testing harden the implementation.

As for 2), agree! We should add either some footer or some checksum to the file to handle cases when a crash occurs during writing. However, if we introduce either a checksum or header, I think we should take the opportunity to define a format for the marker file then:

![image](https://github.com/grafana/agent/assets/2617411/45b6003e-f6b7-4ca6-a18d-be7a69c75b36)

[source](https://app.diagrams.net/?src=about#G1TfBPXEcZY7sgOJXyfdaKZfI9vnfKvD_3)

For the error checking code, we can use [CRC](https://pkg.go.dev/hash/crc32) which is supported in Go. Also, if we were to change to an append only marker store, 12 byte is a not too big size of appended entries. Wdyt?

For 3) I guess we could, but this introduces the problem of the append-only marker to grow and grow, which long term will introduce us problems we need to tackle like marker rotation, or cropping it when it grows too large. If we can, I'll start with a simple swap marker like we have, and then since it's hidden behind the `MarkerHandler` we can rethink it.

Reply:
@ptodev wdyt of https://github.com/grafana/agent/pull/5655?

Reply:
os.Rename isn't atomic with the defaults it uses, you need to use something like https://github.com/natefinch/atomic. Specifically it needs to pass in MOVEFILE_WRITE_THROUGH which golang doesn't. https://cs.opensource.google/go/go/+/master:src/cmd/vendor/golang.org/x/sys/windows/syscall_windows.go?q=windows.rename&ss=go%2Fgo

Reply:
Aha, I think that lib is small enough and with a simple enough API that we can add it. For POSIX it just calls `os.Rename`, and for windows it uses a diff syscall. On it!

Reply:
Agreed

Reply:
Closing this thread with agreement on marker file format and atomic swap, which is merged in https://github.com/grafana/agent/pull/5590/commits/97e87dee9216164c7364acae344f050f8ecbab3c